# **Laboratorio 01: Regresión Lineal - Peticiones Cliente VS Respuesta**
**Estudiante:** Carlos Daniel Kevin Apaza Villca - Oyente

**Tema del Dataset:** Rendimiento y latencia de un servidor web.

**Descripción de las variables:**
* **Variable X (Independiente):** Cantidad de peticiones concurrentes por minuto al servidor.
* **Variable Y (Dependiente):** Tiempo de respuesta promedio del servidor (en milisegundos).

**Lógica:** A medida que la cantidad de peticiones concurrentes ($X$) aumenta, el procesador se satura y el tiempo de respuesta ($Y$) se incrementa de forma proporcional. Buscamos encontrar los mejores parámetros $b$ y $w$ para la ecuación $y = b + wx$.

**Paso 1** - Carga del DataSet

In [ ]:
# PASO 1: Nuestro Dataset de Servidor (40 filas)
# X = Peticiones por minuto -> Es la cantidad de clientes o usuarios intentando entrar a tu servidor al mismo tiempo.
# Y = Tiempo de respuesta en milisegundos (ms) -> Es lo que tarda tu servidor en contestarles a esos usuarios (medido en milisegundos).

X = [10, 15, 20, 22, 30, 35, 40, 45, 50, 52, 60, 65, 70, 75, 80, 85, 90, 95, 100, 105, 110, 115, 120, 125, 130, 135, 140, 145, 150, 155, 160, 165, 170, 175, 180, 185, 190, 195, 200, 210]
Y = [45, 68, 85, 92, 115, 132, 148, 160, 180, 190, 205, 220, 235, 245, 260, 280, 295, 310, 325, 340, 355, 365, 385, 400, 415, 430, 440, 465, 480, 495, 510, 525, 540, 550, 575, 590, 605, 620, 635, 660]

print("Total de datos X cargados:", len(X))
print("Total de datos Y cargados:", len(Y))

Total de datos X cargados: 40
Total de datos Y cargados: 40


**Paso 2** - Entrenamiento (fuerza bruta)

In [ ]:
# PASO 2: Variables para guardar a los "ganadores"
mejor_b = 0
mejor_w = 0
menor_error = 9999999 # Empezamos con un error gigante para que el primer intento lo supere

print("Iniciando el entrenamiento de fuerza bruta para el servidor...")

# PASO 3: Los bucles (Estructuras repetitivas)
# Evaluaremos el sesgo 'b' (latencia base) desde 0 hasta 100 milisegundos
for b in range(0, 100):

    # Evaluaremos el peso 'w' (Como el for no usa decimales, probamos del 0 al 500 y dividimos entre 100)
    # Esto probara pesos desde 0.00 hasta 5.00 milisegundos por petición
    for w_int in range(0, 500):
        w = w_int / 100

        error_total_del_intento = 0

        # Recorremos nuestros 40 datos del servidor
        for i in range(len(X)):
            x_actual = X[i] # Peticiones reales
            y_real = Y[i]   # Tiempo de respuesta real

            # LA FORMULA MATEMATICA -> Nuestra predicción
            y_prediccion = b + (w * x_actual)

            # Calculamos la diferencia absoluta entre la realidad y la prediccion (el Error)
            error = abs(y_real - y_prediccion)
            error_total_del_intento = error_total_del_intento + error

        # Estructura de control decisivo (El condicional)
        # Si esta combinacion de b y w logro un error menor, los guardamos como los mejores
        if error_total_del_intento < menor_error:
            menor_error = error_total_del_intento
            mejor_b = b
            mejor_w = w

print("¡Entrenamiento terminado!")

print("--- RESULTADOS DEL MODELO ---")

print("El mejor Sesgo (b - latencia base) encontrado es:", mejor_b, "ms")
print("El mejor Peso (w - ms extra por peticion) encontrado es:", mejor_w)
print("Margen de error total acumulado:", menor_error)

Iniciando el entrenamiento de fuerza bruta para el servidor...
¡Entrenamiento terminado!
--- RESULTADOS DEL MODELO ---
El mejor Sesgo (b - latencia base) encontrado es: 22 ms
El mejor Peso (w - ms extra por petición) encontrado es: 3.05
Margen de error total acumulado: 119.05000000000001


**Paso 3** - Inferencias y Predicciones

In [ ]:
# PASO 4: Inferencias (Prediciendo el futuro del servidor)
print("\n--- PREDICCIONES DE LATENCIA PARA EL SERVIDOR ---")

# Inventamos 10 cantidades nuevas de peticiones concurrentes (Nuevos valores de X)
# Simularemos desde un trafico alto normal, hasta un ataque masivo de 2,000 peticiones
nuevas_peticiones = [250, 300, 350, 400, 500, 750, 800, 1000, 1500, 2000]

for x_nuevo in nuevas_peticiones:
    # Usamos nuestra formula ya entrenada: y = b + wx
    prediccion_tiempo = mejor_b + (mejor_w * x_nuevo)

    # Redondeamos a 2 decimales para que el tiempo en milisegundos se vea limpio
    tiempo_redondeado = round(prediccion_tiempo, 2)

    print(f"Si entran {x_nuevo} peticiones/min, el servidor tardara aprox: {tiempo_redondeado} ms en responder.")


--- PREDICCIONES DE LATENCIA PARA EL SERVIDOR ---
Si entran 250 peticiones/min, el servidor tardará aprox: 784.5 ms en responder.
Si entran 300 peticiones/min, el servidor tardará aprox: 937.0 ms en responder.
Si entran 350 peticiones/min, el servidor tardará aprox: 1089.5 ms en responder.
Si entran 400 peticiones/min, el servidor tardará aprox: 1242.0 ms en responder.
Si entran 500 peticiones/min, el servidor tardará aprox: 1547.0 ms en responder.
Si entran 750 peticiones/min, el servidor tardará aprox: 2309.5 ms en responder.
Si entran 800 peticiones/min, el servidor tardará aprox: 2462.0 ms en responder.
Si entran 1000 peticiones/min, el servidor tardará aprox: 3072.0 ms en responder.
Si entran 1500 peticiones/min, el servidor tardará aprox: 4597.0 ms en responder.
Si entran 2000 peticiones/min, el servidor tardará aprox: 6122.0 ms en responder.


## **Explicación de los Resultados Obtenidos**


**El Significado de los Parámetros:** La computadora descubrió por sí sola que el mejor sesgo (latencia base) es 22 ms y el mejor peso es 3.05. Llevado a la vida real, esto significa que el servidor, sin estrés, tarda 22 ms en responder, pero por cada petición concurrente extra que entra, el tiempo de respuesta se retrasa 3.05 ms

---

**Un Modelo Preciso:** El margen de error total acumulado de 119.05 nos demuestra que el programa logró trazar una línea de tendencia sumamente precisa para los 40 datos históricos del servidor.

---
**Las 10 Inferencias (Prediciendo el Futuro):** En la última etapa, probamos el modelo con 10 escenarios nuevos de alto tráfico. Los resultados nos permiten anticipar el comportamiento del sistema. Por ejemplo, a 250 peticiones por minuto, el tiempo de respuesta sigue siendo aceptable (784.5 ms). Sin embargo, bajo un ataque o carga extrema de 2,000 peticiones concurrentes, el modelo predice matemáticamente que el servidor tardará 6122.0 ms (más de 6 segundos) en responder, lo que en el mundo del desarrollo web se consideraría una caída del sistema.